### PR TITLE
libct/cg/sd: fix dbus connection leak (alternative)

### PR DESCRIPTION
**Pull request by kolyshkin**
TL;DR: fixes a regression (open fd leak) caused by PR #2923. *This is an alternative to #2936.*

When running libcontainer/integration/TestSystemdFreeze 2000 times,
I got "too many open files" error after a while:

```console
[vagrant@localhost integration]$ sudo -E PATH=$PATH ./integration.test -test.v -test.run Freeze -test.count 2000
<...>
=== RUN   TestFreeze
--- PASS: TestFreeze (0.28s)
=== RUN   TestSystemdFreeze
--- PASS: TestSystemdFreeze (0.42s)
=== RUN   TestFreeze
    utils_test.go:86: exec_test.go:536: unexpected error: container_linux.go:380: starting container process caused: process_linux.go:338: starting init process command caused: fork/exec /proc/self/exe: too many open files
        
--- FAIL: TestFreeze (0.16s)
=== RUN   TestSystemdFreeze
    utils_test.go:86: exec_test.go:536: unexpected error: container_linux.go:380: starting container process caused: process_linux.go:338: starting init process command caused: fork/exec /proc/self/exe: too many open files

<...>

--- FAIL: TestFreeze (0.00s)
=== RUN   TestSystemdFreeze
    utils_test.go:86: exec_test.go:536: unexpected error: container_linux.go:380: starting container process caused: process_linux.go:338: starting init process command caused: open /dev/null: too many open files
        
--- FAIL: TestSystemdFreeze (0.16s)
=== RUN   TestFreeze
    utils_test.go:86: exec_test.go:512: unexpected error: untar error "fork/exec /bin/sh: too many open files": ""
        
--- FAIL: TestFreeze (0.00s)
=== RUN   TestSystemdFreeze
    utils_test.go:86: exec_test.go:536: unexpected error: container_linux.go:364: creating new parent process caused: container_linux.go:496: including execfifo in cmd.Exec setup caused: open /tmp/libcontainer548959440/test/exec.fifo: too many open files
       
 --- FAIL: TestSystemdFreeze (0.17s)
=== RUN   TestFreeze
    utils_test.go:86: exec_test.go:512: unexpected error: untar error "open /dev/null: too many open files": ""
        
<...>
```

Indeed, systemd cgroup manager never closes its dbus connection.
    
This was not a problem before PR #2923 because we only had a single connection
for the whole runtime. Now it is per manager instance, so we leak a
connection (apparently two sockets) per cgroup manager instance.
    
The fix is to go back to using a single global dbus connection.
**UPDATE** this is now done to minimize the diff size.

This also makes it impossible to use both rootful and  rootless dbus connections
at the same time. From what I understand, no runtime ever wants or needs to
do that, so it's not an issue in practice. An assertion is added to make sure this
never happens.

A test case is added in a separate commit. Before the fix, it always fails
like this:
```
        exec_test.go:2030: extra fd 8 -> socket:[659703]
        exec_test.go:2030: extra fd 11 -> socket:[658715]
        exec_test.go:2033: found 2 extra fds after container.Run
    --- FAIL: TestFdLeaksSystemd (0.20s)

```

### CODE REVIEW

File: libcontainer/cgroups/systemd/v2.go [hunk 1]
```diff
@@ -337,12 +335,12 @@ func (m *unifiedManager) getSliceFull() (string, error) {
 	}
 
 	if m.rootless {
-		dbusConnection, err := m.dbus.getConnection()
-		if err != nil {
-			return "", err
-		}
-		// managerCGQuoted is typically "/user.slice/user-${uid}.slice/user@${uid}.service" including the quote symbols
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/systemd/v2.go:344]: Can we retain this comment

Comment (kolyshkin) [reply libcontainer/cgroups/systemd/v2.go:344]: right; fixed

**Comment by kolyshkin** [comment]
Rebased (minor conflicts in libct/int), addressed a review nit.

**Comment by kolyshkin** [comment]
CI failure is a known flake (#2907) which is appearing a lot lately.

**Comment by kolyshkin** [comment]
I have rewritten this PR to be less intrusive (don't change any of the callers). PTAL @cyphar @AkihiroSuda @mrunalp

**Review by mrunalp — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
@kolyshkin Do you want https://github.com/opencontainers/runc/pull/2936 or this one?

**Comment by kolyshkin** [comment]
> Do you want #2936 or this one?

Good question. @cyphar thinks #2936 is better (but he had not seen the last iteration of this one). @mrunalp thinks this one is better. I slightly favor this one -- it is less elegant than #2936 but practically makes more sense (why do we need more than 1 connection anyway?).

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Merging. We can revisit https://github.com/opencontainers/runc/pull/2936 later if it turned out to be better.
